File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/bfs/P752_OpenTheLock.java
package com.dsa.leetcode.labuladong.bfs;

import java.util.*;

public class P752_OpenTheLock {
    public int openLock(String[] deadends, String target) {
        Set<String> visited = new HashSet<>(Arrays.asList(deadends));
        if (visited.contains(target) || visited.contains("0000")) {
            return -1;
        }

        Queue<String> queue = new LinkedList<>();
        // 从0000开始搜索
        queue.add("0000");
        int minStep = 0;

        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                String lockStatus = queue.poll();
                if (target.equals(lockStatus)) {
                    return minStep;
                }
                String[] strings = getNeighbors(lockStatus);
                for (String neighbor : strings) {
                    if (visited.add(neighbor)) {
                        queue.add(neighbor);
                    }
                }
            }
            minStep++;
        }
        return -1;
    }

    private String[] getNeighbors(String lockStatus) {
        String[] neighbors = new String[8];
        int index = 0;

        char[] chars = lockStatus.toCharArray();
        for (int i = 0; i < 4; i++) {
            char temp = chars[i];
            chars[i] = addOne(temp);
            neighbors[index++] = String.valueOf(chars);
            chars[i] = minusOne(temp);
            neighbors[index++] = String.valueOf(chars);
            chars[i] = temp;
        }
        return neighbors;
    }

    private char minusOne(char temp) {
        return temp == '0' ? '9' : (char) (temp - 1);
    }

    private char addOne(char temp) {
        return temp == '9' ? '0' : (char) (temp + 1);
    }

    public static void main(String[] args) {
        String[] deadends = {"0201", "0101", "0102", "1212", "2002"};
        String target = "0202";
        P752_OpenTheLock p752OpenTheLock = new P752_OpenTheLock();
        System.out.println(p752OpenTheLock.openLock(deadends, target));
    }
}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/bfs/P752_OpenTheLock.py
# author: tobebetter9527
# since: 2023/11/20 17:37
from collections import deque
from typing import List


class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        if target == "0000":
            return 0
        dead = set(deadends)
        if "0000" in dead or target in dead:
            return -1
        q = deque()
        q.append("0000")
        step = 0

        while q:
            size = len(q)
            for i in range(size):
                lockStatus = q.popleft()
                if lockStatus == target:
                    return step
                neighbors = self.getNeighbors(lockStatus)
                for neighbor in neighbors:
                    if neighbor not in dead:
                        dead.add(neighbor)
                        q.append(neighbor)
            step += 1
        return -1

    def getNeighbors(self, lockStatus):
        neighbors = []
        chars = list(lockStatus)
        for i in range(4):
            temp = chars[i]
            chars[i] = self.addOne(temp)
            neighbors.append("".join(chars))
            chars[i] = self.minusOne(temp)
            neighbors.append("".join(chars))
            chars[i] = temp
        return neighbors

    def addOne(self, temp):
        return "0" if temp == "9" else str(int(temp) + 1)

    def minusOne(self, temp):
        return "9" if temp == "0" else str(int(temp) - 1)


if __name__ == '__main__':
    deadends = ["8887", "8889", "8878", "8898", "8788", "8988", "7888", "9888"]
    target = "8888"
    solu = Solution()
    print(solu.openLock(deadends, target))
